crates/dojo-lang/src/plugin_test_data/system

```diff
@@ -84,6 +84,134 @@ mod withcomponent {
         testcomponent2_event: testcomponent2::Event,
     }
 }
+#[dojo::interface]
+trait IEmptyTrait;
+
+#[dojo::interface]
+trait IFaultyTrait {
+    /// Even if it looks weird, the expected pattern for a constant in a trait is
+    /// `const <name>: <type>;`
+    /// See: https://github.com/starkware-libs/cairo/blob/600553d34a1af93ff7fb038ac677efdd3fd07010/crates/cairo-lang-parser/src/parser.rs#L835
+    const ONE: u8;
+
+    fn do_ref_self(ref self: TContractState);
+
+    #[my_attr]
+    fn do_with_attrs(p1: u8) -> u16;
+}
+
+#[dojo::interface]
+trait INominalTrait {
+    fn do_no_param();
+    fn do_no_param_but_self(self: @TContractState);
+    fn do_params(p1: dojo_examples::models::Direction, p2: u8);
+    fn do_params_and_self(self: @TContractState, p2: u8);
+    fn do_return_value(p1: u8) -> u16;
+}
+
+#[dojo::contract]
+mod MyFaultyContract {
+
+    fn do_with_ref_self(ref self: ContractState) -> felt252 {
+        'land'
+    }
+
+    fn do_with_several_world_dispatchers(
+        world: IWorldDispatcher,
+        vec: Vec2,
+        another_world: IWorldDispatcher
+    ) -> felt252 {
+        'land'
+    }
+
+    #[external(v0)]
+    impl ActionsImpl of IActions<ContractState> {
+        fn do_with_ref_self(ref self: ContractState) -> felt252 {
+            'land'
+        }
+
+        fn do_with_several_world_dispatchers(
+            world: IWorldDispatcher,
+            vec: Vec2,
+            another_world: IWorldDispatcher
+        ) -> felt252 {
+            'land'
+        }
+    }
+}
+
+#[dojo::contract]
+mod MyNominalContract {
+
+    #[derive(Drop)]
+    struct Action {
+        damage: u8
+    }
+
+    #[external(v0)]
+    #[computed]
+    fn do(vec: Vec2) -> felt252 {
+        'land'
+    }
+
+    #[external(v0)]
+    #[computed]
+    fn do_with_self(self: @ContractState, vec: Vec2) -> felt252 {
+        'land'
+    }
+
+    #[external(v0)]
+    #[computed]
+    fn do_with_world_first(world: IWorldDispatcher, vec: Vec2) -> felt252 {
+        'land'
+    }
+
+    #[external(v0)]
+    #[computed]
+    fn do_with_self_and_world_first(self: @ContractState, world: IWorldDispatcher, vec: Vec2) -> felt252 {
+        'land'
+    }
+
+    #[external(v0)]
+    #[computed]
+    fn do_with_world_not_first(vec: Vec2, world: IWorldDispatcher) -> felt252 {
+        'land'
+    }
+
+    #[external(v0)]
+    #[computed]
+    fn do_with_self_and_world_not_first(self: @ContractState, vec: Vec2, world: IWorldDispatcher) -> felt252 {
+        'land'
+    }
+
+    #[external(v0)]
```

Comment:
`abi(embed_v0)` is a bit more painful to use as it can't be used on standalone function. But `external(v0)` is expected to be deprecated. Would it be possible to switch to `abi(embed_v0)` in this PR?

Reply:
Hum, from what I understand, these attributes are not really important in this `system` file because they are not really used by the plugin to generate some specific code. But it is better to have at least one attribute on each functions to be sure that function attributes are well copied in the output code.

I've updated the code to use `abi(embed_v0)` on `Impl` block and `external(v0)` on free functions. What should we use instead of `external(v0)` for free functions ? Or do we have to put external/public functions in a `Impl` block to have them exported in the ABI file maybe ?